test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,219 @@
+		}
+	}
+	foreach(func(invert bool, i int) {
+		myConfig := rest.CopyConfig(loopbackConfig)
+		myConfig = rest.AddUserAgent(myConfig, fmt.Sprintf("invert=%v, i=%d", invert, i))
+		myClientset := clientset.NewForConfigOrDie(myConfig)
+		fcIfc := myClientset.FlowcontrolV1beta1()
+		fsIfc := fcIfc.FlowSchemas()
+		if lastRVs == nil {
+			lastRVs = make(map[string]string)
+			// Wait until every FlowSchema is defined, and record its RV
+			allFlowSchemas := append(fcboot.MandatoryFlowSchemas, fcboot.SuggestedFlowSchemas...)
+			err := wait.Poll(time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+				for _, fs := range allFlowSchemas {
+					_, err := fsIfc.Get(ctx, fs.Name, metav1.GetOptions{})
+					if err != nil {
+						return false, err
+					}
+					t.Logf("Got initial FlowSchema %#+v", fs)
+					key, _ := cache.MetaNamespaceKeyFunc(fs)
+					lastRVs[key] = fs.ResourceVersion
+				}
+				return true, nil
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+		informerFactory := informers.NewSharedInformerFactory(myClientset, 0)
+		fieldMgr := utilfc.ConfigConsumerAsFieldManager
+		foundToDangling := func(found bool) bool { return !found }
+		if invert {
+			fieldMgr = fieldMgr + "x"
+			foundToDangling = func(found bool) bool { return found }
+		}
+		notifiedRVs[invert][i] = map[string]string{}
+		ctlr := utilfc.NewTestable(utilfc.TestableConfig{
+			Name:  fmt.Sprintf("Controller%d[invert=%v]", i, invert),
+			Clock: clk,
+			FinishHandlingNotification: func(wq workqueue.RateLimitingInterface, obj interface{}) {
+				t.Logf("For invert=%v, i=%v, notified of %#+v", invert, i, obj)
+				obj = peel(obj)
+				switch typed := obj.(type) {
+				case *flowcontrol.FlowSchema:
+					key, _ := cache.MetaNamespaceKeyFunc(obj)
+					notifiedRVs[invert][i][key] = typed.ResourceVersion
+				}
+			},
+			AsFieldManager:         fieldMgr,
+			FoundToDangling:        foundToDangling,
+			InformerFactory:        informerFactory,
+			FlowcontrolClient:      fcIfc,
+			ServerConcurrencyLimit: 200,         // server concurrency limit
+			RequestWaitLimit:       time.Minute, // request wait limit
+			ObsPairGenerator:       metrics.PriorityLevelConcurrencyObserverPairGenerator,
+			QueueSetFactory:        fqtesting.NewNoRestraintFactory(),
+		})
+		ctlrs[invert][i] = ctlr
+		informerFactory.Start(stopCh)
+		if ctlr.WaitForCacheSync(stopCh) {
+			t.Logf("Achieved initial sync for invert=%v, i=%v", invert, i)
+		} else {
+			t.Fatalf("Never achieved initial sync for invert=%v, i=%v", invert, i)
+		}
+	})
+	someTimedOut := false
+	var writeCount, countAtStart int
+	var tStart time.Time
+	const testN = 30
+	nextTime := clk.Now()
+	for j := 0; j < 2+testN; j++ {
+		AOK := false
+		// wait until notifiedRVs[invert][i] covers lastRVs for all invert, i
+		for k := 0; k < 100 && !AOK; k++ {
+			time.Sleep(time.Millisecond * 100)
+			AOK = true
+			foreach(func(invert bool, i int) {
+				for key, rv := range lastRVs {
+					if notifiedRVs[invert][i][key] != rv {
```

Comment:
With this written on line 118, how do you avoid the data race?

Reply:
Good point, the writes are in different goroutines.

Reply:
Added synchronization.